# Pre-Procesamiento de datos: Desercion Universidad de Boyaca

# 1. Cargando tablas e informacion general

In [1]:
# importar pandas
import pandas as pd

In [2]:
docentes_periodo_df = pd.read_csv('../data/raw/docentes_periodo.csv', ';', encoding='ISO-8859-1')
eval_docentes_periodo_df = pd.read_csv('../data/raw/eval_docentes_periodo.csv', ';', encoding='ISO-8859-1')
mat_cursadas_df = pd.read_csv('../data/raw/mat_cursadas_rend_academico.csv', ';', encoding='ISO-8859-1')
perfil_ingreso_df = pd.read_csv('../data/raw/perfil_ingreso.csv', ';')
rendimiento_academico_df = pd.read_csv('../data/raw/rendimiento_academico.csv', ';', encoding='ISO-8859-1')

area_icfes_df = pd.read_excel('../data/raw/area_icfes.xlsx')
creditos_periodo_df = pd.read_excel('../data/raw/CREDITOS_PERIODO.xlsx')
perfil_ingreso_v2_df = pd.read_excel('../data/raw/perfil_ingreso_v2.xlsx')
materias_por_programa_df = pd.read_excel('../data/raw/materias_por_programa.xlsx')
pre_correquisito_df = pd.read_excel('../data/raw/pre_correquisito.xlsx')
mat_cursadas_v2_df = pd.read_excel('../data/raw/mat_cursadas_rend_academico_v2.xlsx')
reingreso_aplazamiento_df = pd.read_excel('../data/raw/reingreso_aplazamiento.xlsx')

In [3]:
# vistazo de las tablas
print("\n"+"docentes_periodo".center(90))
display(docentes_periodo_df.head(3))
print("\n"+"eval_docentes_periodo".center(90))
display(eval_docentes_periodo_df.head(3))
print("\n"+"mat_cursadas".center(90))
display(mat_cursadas_df.head(3))
print("\n"+"perfil_ingreso_df".center(90))
display(perfil_ingreso_df.head(3))
print("\n"+"rendimiento_academico".center(90))
display(rendimiento_academico_df.head(3))

print("\n"+"area_icfes".center(90))
display(area_icfes_df.head(3))
print("\n"+"creditos_periodo".center(90))
display(creditos_periodo_df.head(3))
print("\n"+"perfil_ingreso_v2".center(90))
display(perfil_ingreso_v2_df.head(3))
print("\n"+"materias_por_programa".center(90))
display(materias_por_programa_df.head(3))
print("\n"+"pre_correquisito".center(90))
display(pre_correquisito_df.head(3))
print("\n"+"mat_cursadas_v2".center(90))
display(mat_cursadas_v2_df.head(3))
print("\n"+"reingreso_aplazamiento".center(90))
display(reingreso_aplazamiento_df.head(3))


                                     docentes_periodo                                     


PERIODO  COD_MAT  COD_DOCENTE
0   201210    11303         1329
1   201210    11304         3111
2   201210    11305          622


                                  eval_docentes_periodo                                   


PERIODO  COD_DOCENTE  COD_MAT  COD_FACTOR  \
0   201210         1008    66410           1   
1   201210         1008    66410           2   
2   201210         1008    66410           3   

                             NOM_FACTOR CALIFICACION  
0              Asistencia y Puntualidad         4,63  
1              Relación profesor alumno         4,65  
2  Orientación del proceso de formación         4,58


                                       mat_cursadas                                       


PERIODO_COHORTE  PERIODO_ACADEMICO    CODIGO  TOT_MAT_INSCRITAS  \
0           201210             201210  11312001                  8   
1           201210             201210  11312002                  8   
2           201210             201210  11312003                  8   

   MAT_APROBADAS  
0              7  
1              7  
2              7


                                    perfil_ingreso_df                                     


PERIODO_COHORTE    CODIGO GENERO  EDAD  \
0           201210  11312001      M  22.0   
1           201210  11312002      M  23.0   
2           201210  11312003      F  22.0   

                                 COLEGIO_PROCEDENCIA  PUNT_TOTAL  PUNT_AREA  \
0                              COL COOP COOSERVICIOS         438         49   
1  INSTITUTO DE EDUCACION MEDIA DIVERSIFICADA INE...         386         48   
2                                I.E. ANTONIO NARI�O         460         51   

  ES_DESERTOR  
0          NO  
1          NO  
2          NO


                                  rendimiento_academico                                   


PERIODO_COHORTE  PERIODO_ACADEMICO    CODIGO  COD_MAT         NOMBRE_MAT  \
0           201210             201210  11312001    11305  EXP. ORAL ESCRITA   
1           201210             201210  11312001    11312  INICIATIVA EMPRES   
2           201210             201210  11312001    11310      MICROECONOMÍA   

  NOTA_DEF    ESTADO  
0     3,50  APROBADA  
1     4,00  APROBADA  
2     3,60  APROBADA


                                        area_icfes                                        


FECHA_INI_EX FECHA_FIN_EX  COD_MEN          AREA_ICFES
0   1980-05-01   1999-12-31    20697  CIENCIAS NATURALES
1   1980-05-01   1999-12-31    20697            LENGUAJE
2   1980-05-01   1999-12-31    20697          MATEMÁTICA


                                     creditos_periodo                                     


CODIGO  PERIODO
0  33111541   201210
1  66411591   201210
2  33103565   201210


                                    perfil_ingreso_v2                                     


PERIODO_COHORTE    CODIGO GENERO  EDAD  \
0           201210  11312001      M  22.0   
1           201210  11312002      M  23.0   
2           201210  11312003      F  22.0   

                                 COLEGIO_PROCEDENCIA  PUNT_TOTAL  PUNT_AREA  \
0                              COL COOP COOSERVICIOS         438         49   
1  INSTITUTO DE EDUCACION MEDIA DIVERSIFICADA INE...         386         48   
2                                I.E. ANTONIO NARIÑO         460         51   

  ES_DESERTOR  ULTIMA_MATRICULA  COD_MEN  
0          NO            201710    20708  
1          NO            201420    20708  
2          NO            201620    20708


                                  materias_por_programa                                   


COD_MEN                                          FACULTAD  \
0    20708  FACULTAD DE CIENCIAS ADMINISTRATIVAS Y CONTABLES   
1    20708  FACULTAD DE CIENCIAS ADMINISTRATIVAS Y CONTABLES   
2    20708  FACULTAD DE CIENCIAS ADMINISTRATIVAS Y CONTABLES   

              NOMBRE_PROGRAMA ACREDITADO        SEDE  COD_MAT  NUM_CREDITOS  \
0  ADMINISTRACIÓN DE EMPRESAS       N.A.  SEDE TUNJA    11330             2   
1  ADMINISTRACIÓN DE EMPRESAS       N.A.  SEDE TUNJA    11329             2   
2  ADMINISTRACIÓN DE EMPRESAS       N.A.  SEDE TUNJA    11334             2   

  INTRODUCCION  
0           NO  
1           NO  
2           NO


                                     pre_correquisito                                     


COD_BASE              NOMBRE_MAT  COD_PRECO NOMBRE_PRECORREQUISITO  \
0     11304    PLANEACIÓN Y CONTROL      11313  ADMINISTRACIÓN BÁSICA   
1     11308     CÁLCULO DIFERENCIAL      11301      MATEMÁTICA BÁSICA   
2     11309  CONTABILIDAD COMERCIAL      11311    CONTABILIDAD BÁSICA   

  TIPO_REQUISITO  
0  PRERREQUISITO  
1  PRERREQUISITO  
2  PRERREQUISITO


                                     mat_cursadas_v2                                      


PERIODO_COHORTE  PERIODO_ACADEMICO    CODIGO  TOT_MAT_INSCRITAS  \
0           201210             201210  11312001                  8   
1           201210             201210  11312002                  8   
2           201210             201210  11312003                  8   

   CRED_INSCRITOS  MAT_APROBADAS  CRED_APROBADOS  
0            17.0              7            15.0  
1            16.0              7            14.0  
2            17.0              7            15.0


                                  reingreso_aplazamiento                                  


CODIGO FECHA_SOLICITUD       TIPO
0  55110503      2012-01-10  REINGRESO
1  55510048      2012-01-10  REINGRESO
2  55510535      2012-01-10  REINGRESO

# 2. Tabla "docentes_periodo"

In [4]:
docentes_periodo_df.head()

PERIODO  COD_MAT  COD_DOCENTE
0   201210    11303         1329
1   201210    11304         3111
2   201210    11305          622
3   201210    11312         1329
4   201210    11313         1329

In [5]:
docentes_periodo_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32537 entries, 0 to 32536
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   PERIODO      32537 non-null  int64
 1   COD_MAT      32537 non-null  int64
 2   COD_DOCENTE  32537 non-null  int64
dtypes: int64(3)
memory usage: 762.7 KB


In [6]:
docentes_periodo_df.nunique()

PERIODO          37
COD_MAT        2275
COD_DOCENTE    2220
dtype: int64

> * las 3 columnas deben ser variables categoricas no numericas, asi que debemos convertirlos a strings
> * NO hay valores faltantes


## 2.1 Llenar valores faltantes

> NO hay valores faltantes

## 2.2 Corregir Tipos de datos:

Convertimos las variables numericas que son categoricas a string

In [7]:
docentes_periodo_df = docentes_periodo_df.astype({'PERIODO':str, 'COD_MAT':str, 'COD_DOCENTE':str})

## 2.3 Normalizar Texto

> No hay campos de texto

## 2.4 Revisar Llaves primarias de Categorias

> No hay categorias a relacionar con las llaves

## 2.5 Expandir datos

* ### Agregar columna año y semestre:

    A partir de la columna "PERIODO" creamos las columnas "AÑO" y "SEMESTRE

In [8]:
def expandir_periodo(serie):
    año = serie.str.slice(0,4)
    semestre = serie.str.slice(4).map({'10':'semestre 1',
                                        '20':'semestre 2',
                                        '11':'intersemestral 1',
                                        '21':'intersemestral 2'})
        
    return año, semestre

In [9]:
docentes_periodo_df['AÑO'], docentes_periodo_df['SEMESTRE'] = expandir_periodo(docentes_periodo_df.PERIODO)

* ### Agregar columna fecha (en caso de que debamos graficar la data comom una serie temporal):

In [10]:
def fecha_de_semestre(años, semestres):
    fecha = años + '-' + semestres.map({  'semestre 1':'02',
                                                'intersemestral 1':'06',
                                                'semestre 2':'08',
                                                'intersemestral 2':'12'})
    fecha = pd.to_datetime(fecha)
    return fecha

In [11]:
# docentes_periodo_df['FECHA'] = fecha_de_semestre(docentes_periodo_df['AÑO'], docentes_periodo_df['SEMESTRE'])

## 2.6 Consistencia en los datos

Parece que:

* los valores de "COD_DOCENTE" deben ser de 4 digitos
* los valores de "PERIODO" deben ser de 6 digitos
* los valores de 

Verificamos:

In [12]:
docentes_periodo_df.PERIODO.str.len().value_counts()

6    32537
Name: PERIODO, dtype: int64

> * Todos tienen exactamente 6 digitos

In [13]:
docentes_periodo_df.COD_DOCENTE.str.len().value_counts()

4    29133
3     2841
2      448
1      115
Name: COD_DOCENTE, dtype: int64

> La mayoria tienen 4 digitos, algunos tienen menos

In [14]:
docentes_periodo_df.COD_MAT.str.len().value_counts()

5    32371
6      166
Name: COD_MAT, dtype: int64

> La mayoria tienen 5 digitos, unos pocos tienen mas (exactamente 6 digitos)

Rellenamos con ceros a la izquierda para tener consistencia en cuanto a el numero de digitos. Pero, solo lo hacemos si no modificamos la cantidad de valores unicos.

In [15]:
def fill_digits(serie):
    ndigits = serie.str.len().max()
    filled = serie.str.zfill(ndigits)
    if serie.nunique() == filled.nunique():
        return filled
    else:
        return serie

In [16]:
#docentes_periodo_df.COD_DOCENTE = fill_digits(docentes_periodo_df.COD_DOCENTE)
#docentes_periodo_df.COD_MAT = fill_digits(docentes_periodo_df.COD_MAT)

## 2.7 Resultado

In [17]:
docentes_periodo_df.head()

PERIODO COD_MAT COD_DOCENTE   AÑO    SEMESTRE
0  201210   11303        1329  2012  semestre 1
1  201210   11304        3111  2012  semestre 1
2  201210   11305         622  2012  semestre 1
3  201210   11312        1329  2012  semestre 1
4  201210   11313        1329  2012  semestre 1

In [18]:
docentes_periodo_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32537 entries, 0 to 32536
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   PERIODO      32537 non-null  object
 1   COD_MAT      32537 non-null  object
 2   COD_DOCENTE  32537 non-null  object
 3   AÑO          32537 non-null  object
 4   SEMESTRE     32537 non-null  object
dtypes: object(5)
memory usage: 1.2+ MB


In [19]:
docentes_periodo_df.nunique()

PERIODO          37
COD_MAT        2275
COD_DOCENTE    2220
AÑO              10
SEMESTRE          4
dtype: int64

In [20]:
docentes_periodo_df.PERIODO.str.len().value_counts()

6    32537
Name: PERIODO, dtype: int64

In [21]:
docentes_periodo_df.COD_DOCENTE.str.len().value_counts()

4    29133
3     2841
2      448
1      115
Name: COD_DOCENTE, dtype: int64

In [22]:
docentes_periodo_df.COD_MAT.str.len().value_counts()

5    32371
6      166
Name: COD_MAT, dtype: int64

## 3. Tabla "eval_docentes_periodo"

In [23]:
eval_docentes_periodo_df.head()

PERIODO  COD_DOCENTE  COD_MAT  COD_FACTOR  \
0   201210         1008    66410           1   
1   201210         1008    66410           2   
2   201210         1008    66410           3   
3   201210         1008    66410           4   
4   201210         1008    66410           5   

                             NOM_FACTOR CALIFICACION  
0              Asistencia y Puntualidad         4,63  
1              Relación profesor alumno         4,65  
2  Orientación del proceso de formación         4,58  
3                         Conocimientos         4,71  
4   Planeación del proceso de formación         4,72

In [24]:
eval_docentes_periodo_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 253035 entries, 0 to 253034
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   PERIODO       253035 non-null  int64 
 1   COD_DOCENTE   253035 non-null  int64 
 2   COD_MAT       253035 non-null  int64 
 3   COD_FACTOR    253035 non-null  int64 
 4   NOM_FACTOR    253035 non-null  object
 5   CALIFICACION  253035 non-null  object
dtypes: int64(4), object(2)
memory usage: 11.6+ MB


In [25]:
eval_docentes_periodo_df.nunique()

PERIODO           18
COD_DOCENTE     2163
COD_MAT         2097
COD_FACTOR        18
NOM_FACTOR        12
CALIFICACION     322
dtype: int64

> * 6 columnas
> * 253035 filas
> * Para esta tabla no hay valores faltantes (nulos)
> * "CALIFICACION" debe ser de tipo float
> * Las demas son varaibles categoricas, las preferimos como strings
> * Hay una inconsistencia en la cantidad de "COD_FACTOR" y "NOM_FACTOR", los cuales deberian ser iguales

## 3.1 Llenar valores Faltantes

> No hay valores faltantes

## 3.2 Corregir tipo de datos

In [26]:
eval_docentes_periodo_df = eval_docentes_periodo_df.astype({'PERIODO':str, 'COD_DOCENTE':str, 'COD_MAT':str, 'COD_FACTOR':str, 'NOM_FACTOR':str})
eval_docentes_periodo_df.CALIFICACION = eval_docentes_periodo_df.CALIFICACION.str.replace(',', '.').astype(float)

## 3.3 Normalizar Texto

In [27]:
def normalize_text(serie):
    # to ascii:
    serie = serie.str.normalize('NFKD')
    serie = serie.str.encode("ascii","ignore")
    serie = serie.str.decode('ascii')
    # to lowercase:
    serie = serie.str.lower()
    return serie

In [28]:
eval_docentes_periodo_df.NOM_FACTOR = normalize_text(eval_docentes_periodo_df.NOM_FACTOR)

## 3.4 Revisar Llaves primarias de Categorias

* ### Corregir "COD_FACTOR" y "NOM_FACTOR" 

In [29]:
eval_docentes_periodo_df.groupby(['COD_FACTOR', 'NOM_FACTOR'])[['COD_DOCENTE']].count()

COD_DOCENTE
COD_FACTOR NOM_FACTOR                                       
1          asistencia y puntualidad                    26947
10         material didactico                           1173
11         tutoria virtual                              1173
12         relacion profesor alumno                     1173
13         orientacion del proceso de formacion         1173
14         conocimientos                                1173
15         planeacion del proceso de formacion          1173
16         evaluacion del aprendizaje                   1173
17         investigacion y extension                    1173
18         compromiso institucional                     1173
2          relacion profesor alumno                    26945
3          orientacion del proceso de formacion        26945
4          conocimientos                               26944
5          planeacion del proceso de formacion         26941
6          evaluacion del aprendizaje                  26939
7          imagen y orden                              26941
8          compromiso institucional                    26937
9          investigacion y extension                   26939

> El problema es que existe mas de un codigo para el mismo factor.

In [30]:
factores = eval_docentes_periodo_df.NOM_FACTOR.unique()
cod_factores = {factor:str(i+1) for i, factor in enumerate(factores)}
cod_factores

{'asistencia y puntualidad': '1',
 'relacion profesor alumno': '2',
 'orientacion del proceso de formacion': '3',
 'conocimientos': '4',
 'planeacion del proceso de formacion': '5',
 'evaluacion del aprendizaje': '6',
 'imagen y orden': '7',
 'compromiso institucional': '8',
 'investigacion y extension': '9',
 'material didactico': '10',
 'tutoria virtual': '11'}

In [31]:
def fix_factor_codes(col):
    factores = col.unique()
    cod_factores = {factor:str(i+1) for i, factor in enumerate(factores)}
    return col.map(cod_factores)

In [32]:
eval_docentes_periodo_df['COD_FACTOR'] = fix_factor_codes(eval_docentes_periodo_df.NOM_FACTOR)

## 3.5 Expandir datos

Añadir año y semestre:

In [33]:
eval_docentes_periodo_df['AÑO'], eval_docentes_periodo_df['SEMESTRE'] = expandir_periodo(eval_docentes_periodo_df.PERIODO)

#eval_docentes_periodo_df['FECHA'] = fecha_de_semestre(eval_docentes_periodo_df['AÑO'], eval_docentes_periodo_df['SEMESTRE'])

## 3.6 Consistencia en los datos

Aplica el mismo analisis de la tabla "docentes_periodo" en el numeral 2.6

In [34]:
# filled = eval_docentes_periodo_df[['PERIODO', 'COD_DOCENTE', 'COD_MAT']].apply(fill_digits, axis=0)
# eval_docentes_periodo_df[['PERIODO', 'COD_DOCENTE', 'COD_MAT']] = filled

## 3.7 Resultado

In [35]:
eval_docentes_periodo_df.head()

PERIODO COD_DOCENTE COD_MAT COD_FACTOR  \
0  201210        1008   66410          1   
1  201210        1008   66410          2   
2  201210        1008   66410          3   
3  201210        1008   66410          4   
4  201210        1008   66410          5   

                             NOM_FACTOR  CALIFICACION   AÑO    SEMESTRE  
0              asistencia y puntualidad          4.63  2012  semestre 1  
1              relacion profesor alumno          4.65  2012  semestre 1  
2  orientacion del proceso de formacion          4.58  2012  semestre 1  
3                         conocimientos          4.71  2012  semestre 1  
4   planeacion del proceso de formacion          4.72  2012  semestre 1

In [36]:
eval_docentes_periodo_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 253035 entries, 0 to 253034
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   PERIODO       253035 non-null  object 
 1   COD_DOCENTE   253035 non-null  object 
 2   COD_MAT       253035 non-null  object 
 3   COD_FACTOR    253035 non-null  object 
 4   NOM_FACTOR    253035 non-null  object 
 5   CALIFICACION  253035 non-null  float64
 6   AÑO           253035 non-null  object 
 7   SEMESTRE      253035 non-null  object 
dtypes: float64(1), object(7)
memory usage: 15.4+ MB


In [37]:
eval_docentes_periodo_df.nunique()

PERIODO           18
COD_DOCENTE     2163
COD_MAT         2097
COD_FACTOR        11
NOM_FACTOR        11
CALIFICACION     322
AÑO                9
SEMESTRE           2
dtype: int64

In [38]:
eval_docentes_periodo_df.groupby(['COD_FACTOR', 'NOM_FACTOR'])[['COD_DOCENTE']].count()

COD_DOCENTE
COD_FACTOR NOM_FACTOR                                       
1          asistencia y puntualidad                    26947
10         material didactico                           1173
11         tutoria virtual                              1173
2          relacion profesor alumno                    28118
3          orientacion del proceso de formacion        28118
4          conocimientos                               28117
5          planeacion del proceso de formacion         28114
6          evaluacion del aprendizaje                  28112
7          imagen y orden                              26941
8          compromiso institucional                    28110
9          investigacion y extension                   28112

## 4. Tabla "mat_cursadas_v2"

In [39]:
mat_cursadas_v2_df.head()

PERIODO_COHORTE  PERIODO_ACADEMICO    CODIGO  TOT_MAT_INSCRITAS  \
0           201210             201210  11312001                  8   
1           201210             201210  11312002                  8   
2           201210             201210  11312003                  8   
3           201210             201210  11312004                  8   
4           201210             201210  11312005                  8   

   CRED_INSCRITOS  MAT_APROBADAS  CRED_APROBADOS  
0            17.0              7            15.0  
1            16.0              7            14.0  
2            17.0              7            15.0  
3            17.0              7            15.0  
4            17.0              7            15.0

In [40]:
mat_cursadas_v2_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61074 entries, 0 to 61073
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   PERIODO_COHORTE    61074 non-null  int64  
 1   PERIODO_ACADEMICO  61074 non-null  int64  
 2   CODIGO             61074 non-null  int64  
 3   TOT_MAT_INSCRITAS  61074 non-null  int64  
 4   CRED_INSCRITOS     60052 non-null  float64
 5   MAT_APROBADAS      61074 non-null  int64  
 6   CRED_APROBADOS     57591 non-null  float64
dtypes: float64(2), int64(5)
memory usage: 3.3 MB


In [41]:
mat_cursadas_v2_df.nunique()

PERIODO_COHORTE         19
PERIODO_ACADEMICO       36
CODIGO               10399
TOT_MAT_INSCRITAS       13
CRED_INSCRITOS          31
MAT_APROBADAS           13
CRED_APROBADOS          30
dtype: int64

> * columnas "CRED_" deben ser enteros
> * hay valores faltantes en "CRED_INSCRITOS" y "CRED_APROBADOS"
> * cual es la diferencia de "PERIODO_COHORTE" vs "PERIODO_ACADEMICO"

* ### "PERIODO_COHORTE" vs "PERIODO_ACADEMICO"

In [42]:
periodo_dif = mat_cursadas_v2_df[mat_cursadas_v2_df.PERIODO_ACADEMICO != mat_cursadas_v2_df.PERIODO_COHORTE]
#(periodo_dif.PERIODO_ACADEMICO + ' - ' + periodo_dif.PERIODO_COHORTE).value_counts()
periodo_dif.PERIODO_COHORTE.astype(str).str.slice(4).value_counts()

10    30624
20    20051
Name: PERIODO_COHORTE, dtype: int64

In [43]:
periodo_dif.PERIODO_ACADEMICO.astype(str).str.slice(4).value_counts()

10    25498
20    24449
11      487
21      241
Name: PERIODO_ACADEMICO, dtype: int64

> * "PERIODO_COHORTE" solo incluye primer semestre y segundo semestre
> * "PERIODO_ACADEMICO" incluyes semestres e intersemestrales

> Se podria manejar PERIODO_COHORTE como "mitad_del año" con valores 1 o 2

## 4.1 Llenar valores faltantes

In [44]:
nan_mask_CRED_INSCRITOS = mat_cursadas_v2_df.CRED_INSCRITOS.isna()
mat_cursadas_v2_df[nan_mask_CRED_INSCRITOS].head()

PERIODO_COHORTE  PERIODO_ACADEMICO    CODIGO  TOT_MAT_INSCRITAS  \
20            201210             201210  11412007                  0   
28            201210             201210  11412018                  0   
34            201210             201210  11412026                  0   
48            201210             201210  11712017                  0   
100           201210             201210  33212008                  0   

     CRED_INSCRITOS  MAT_APROBADAS  CRED_APROBADOS  
20              NaN              0             NaN  
28              NaN              0             NaN  
34              NaN              0             NaN  
48              NaN              0             NaN  
100             NaN              0             NaN

> * parece que los valores nulos para "CRED_INSCRITOS" corresponden a semestres en el que el estudiante no inscribio materias ("TOT_MATINSCRITAS" == 0)

In [45]:
mat_cursadas_v2_df[nan_mask_CRED_INSCRITOS].nunique()

PERIODO_COHORTE       19
PERIODO_ACADEMICO     31
CODIGO               971
TOT_MAT_INSCRITAS      1
CRED_INSCRITOS         0
MAT_APROBADAS          1
CRED_APROBADOS         0
dtype: int64

> * Confirmado, en las filas con "CRED_INSCRITOS" == NaN la columna "TOT_MAT_INSCRITAS" solo toma 1 valor (cero)
> * Podemos entoces reemplazar los valores faltante por ceros
> * podemos hacer lo mismo para "CRED_APROBADOS" (al menos para estas filas)

In [46]:
mat_cursadas_v2_df[nan_mask_CRED_INSCRITOS] = mat_cursadas_v2_df[nan_mask_CRED_INSCRITOS].fillna(0)

> * Aun hay valores faltantes para la columna "CRED_APROBADOS"

In [47]:
nan_mask_CRED_APROBADOS = mat_cursadas_v2_df.CRED_APROBADOS.isna()
mat_cursadas_v2_df[nan_mask_CRED_APROBADOS].head()

PERIODO_COHORTE  PERIODO_ACADEMICO    CODIGO  TOT_MAT_INSCRITAS  \
41            201210             201210  11712009                  7   
50            201210             201210  11712021                  8   
114           201210             201210  33312029                  7   
157           201210             201210  55212009                  8   
177           201210             201210  55312022                  8   

     CRED_INSCRITOS  MAT_APROBADAS  CRED_APROBADOS  
41             14.0              0             NaN  
50             16.0              0             NaN  
114            15.0              0             NaN  
157            16.0              0             NaN  
177            16.0              0             NaN

> * Parece ser que el estudiante no aprobo ninguna materia en ese semestre

In [48]:
mat_cursadas_v2_df[nan_mask_CRED_APROBADOS].nunique()

PERIODO_COHORTE        19
PERIODO_ACADEMICO      32
CODIGO               2212
TOT_MAT_INSCRITAS      11
CRED_INSCRITOS         25
MAT_APROBADAS           1
CRED_APROBADOS          0
dtype: int64

> * Confirmado, en las filas con valores flatantes para "CRED_APROBADOS", "MAT_APROBADAS" solo toma un valor (cero)

In [49]:
mat_cursadas_v2_df[nan_mask_CRED_APROBADOS] = mat_cursadas_v2_df[nan_mask_CRED_APROBADOS].fillna(0)

## 4.2 Corregir tipo de datos

In [50]:
mat_cursadas_types = {'PERIODO_COHORTE':str,'PERIODO_ACADEMICO':str,'CODIGO':str,'TOT_MAT_INSCRITAS':int,'CRED_INSCRITOS':int,'MAT_APROBADAS':int,'CRED_APROBADOS':int}

mat_cursadas_v2_df = mat_cursadas_v2_df.astype(dtype=mat_cursadas_types)

## 4.3 Normalizar Texto

> No hay variables de texto

## 4.4 Revisar llaves primarias de categorias

> No tenemos categorias relacionadas a llaves

## 4.5 Expandir datos

In [51]:
# mat_cursadas_v2_df['AÑO'], mat_cursadas_v2_df['SEMESTRE'] = expandir_periodo(mat_cursadas_v2_df.PERIODO_ACADEMICO)
# mat_cursadas_v2_df['FECHA'] = fecha_de_semestre(mat_cursadas_v2_df['AÑO'], mat_cursadas_v2_df['SEMESTRE'])

## 4.6 Consistencia en los datos

In [52]:
# cols_to_fill = ['PERIODO_COHORTE', 'PERIODO_ACADEMICO', 'CODIGO']
# filled = mat_cursadas_v2_df[cols_to_fill].apply(fill_digits, axis=0)
# mat_cursadas_v2_df[cols_to_fill] = filled

## 4.7 Resultado

In [53]:
mat_cursadas_v2_df.head()

PERIODO_COHORTE PERIODO_ACADEMICO    CODIGO  TOT_MAT_INSCRITAS  \
0          201210            201210  11312001                  8   
1          201210            201210  11312002                  8   
2          201210            201210  11312003                  8   
3          201210            201210  11312004                  8   
4          201210            201210  11312005                  8   

   CRED_INSCRITOS  MAT_APROBADAS  CRED_APROBADOS  
0              17              7              15  
1              16              7              14  
2              17              7              15  
3              17              7              15  
4              17              7              15

In [54]:
mat_cursadas_v2_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61074 entries, 0 to 61073
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   PERIODO_COHORTE    61074 non-null  object
 1   PERIODO_ACADEMICO  61074 non-null  object
 2   CODIGO             61074 non-null  object
 3   TOT_MAT_INSCRITAS  61074 non-null  int32 
 4   CRED_INSCRITOS     61074 non-null  int32 
 5   MAT_APROBADAS      61074 non-null  int32 
 6   CRED_APROBADOS     61074 non-null  int32 
dtypes: int32(4), object(3)
memory usage: 2.3+ MB


## 5. Tabla "perfil_ingreso_v2"

In [55]:
perfil_ingreso_v2_df.head()

PERIODO_COHORTE    CODIGO GENERO  EDAD  \
0           201210  11312001      M  22.0   
1           201210  11312002      M  23.0   
2           201210  11312003      F  22.0   
3           201210  11312004      F  25.0   
4           201210  11312005      M  25.0   

                                 COLEGIO_PROCEDENCIA  PUNT_TOTAL  PUNT_AREA  \
0                              COL COOP COOSERVICIOS         438         49   
1  INSTITUTO DE EDUCACION MEDIA DIVERSIFICADA INE...         386         48   
2                                I.E. ANTONIO NARIÑO         460         51   
3                         I.E. LUIS HERNANDEZ VARGAS          41         42   
4                     IE TEC JOSE IGNACIO DE MARQUEZ          50         49   

  ES_DESERTOR  ULTIMA_MATRICULA  COD_MEN  
0          NO            201710    20708  
1          NO            201420    20708  
2          NO            201620    20708  
3          NO            201620    20708  
4          NO            201610    20708

In [56]:
perfil_ingreso_v2_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10400 entries, 0 to 10399
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   PERIODO_COHORTE      10400 non-null  int64  
 1   CODIGO               10400 non-null  int64  
 2   GENERO               10400 non-null  object 
 3   EDAD                 10398 non-null  float64
 4   COLEGIO_PROCEDENCIA  10400 non-null  object 
 5   PUNT_TOTAL           10400 non-null  int64  
 6   PUNT_AREA            10400 non-null  int64  
 7   ES_DESERTOR          10400 non-null  object 
 8   ULTIMA_MATRICULA     10400 non-null  int64  
 9   COD_MEN              10400 non-null  int64  
dtypes: float64(1), int64(6), object(3)
memory usage: 812.6+ KB


In [57]:
perfil_ingreso_v2_df.nunique()

PERIODO_COHORTE           19
CODIGO                 10400
GENERO                     2
EDAD                      47
COLEGIO_PROCEDENCIA     1898
PUNT_TOTAL               390
PUNT_AREA                249
ES_DESERTOR                2
ULTIMA_MATRICULA          31
COD_MEN                   23
dtype: int64

> * En la columna "EDAD" hay 2 valores faltantes
> * la columna edad deberia ser un entero
> * las columnas de codigo o perido son categoricas, las preferimos como strings

## 5.1 Llenar Valores faltantes

In [58]:
perfil_ingreso_v2_df[perfil_ingreso_v2_df.EDAD.isna()]

PERIODO_COHORTE    CODIGO GENERO  EDAD  \
1577           201310  55713038      F   NaN   
1912           201310  77313007      F   NaN   

                                    COLEGIO_PROCEDENCIA  PUNT_TOTAL  \
1577  INSTITUCIÓN EDUCATIVA TÉCNICO INDUSTRIAL GUSTA...          51   
1912         COLEGIO NUESTRA SEÑORA DEL ROSARIO - TUNJA          65   

      PUNT_AREA ES_DESERTOR  ULTIMA_MATRICULA  COD_MEN  
1577         54          NO            201920    53841  
1912         68          NO            201310    90577

In [59]:
print(f'mediana de edad: {round(perfil_ingreso_v2_df.EDAD.median())}')
print(f'media de edad: {round(perfil_ingreso_v2_df.EDAD.mean())}')
print(f'moda de edad: {round(perfil_ingreso_v2_df.EDAD.mode()[0])}')

mediana de edad: 23
media de edad: 24
moda de edad: 22


> * Como son solo dos valores y no afectara el analisis de manera relevante, podemos rellenar con la moda

In [60]:
perfil_ingreso_v2_df.EDAD = perfil_ingreso_v2_df.EDAD.fillna(22)

## 5.2 Corregir tipo de datos

In [61]:
to_str_cols = ['PERIODO_COHORTE', 'CODIGO', 'ULTIMA_MATRICULA', 'COD_MEN']
perfil_ingreso_v2_df[to_str_cols] = perfil_ingreso_v2_df[to_str_cols].astype(str)

to_int_cols = ['EDAD', 'PUNT_TOTAL', 'PUNT_AREA']
perfil_ingreso_v2_df[to_int_cols] = perfil_ingreso_v2_df[to_int_cols].astype(int)

## 5.3 Normalizar Texto

In [62]:
perfil_ingreso_v2_df.COLEGIO_PROCEDENCIA = normalize_text(perfil_ingreso_v2_df.COLEGIO_PROCEDENCIA)
perfil_ingreso_v2_df.ES_DESERTOR = normalize_text(perfil_ingreso_v2_df.ES_DESERTOR)

## 5.4 Revisar Llaves primarias de categorias

> No hay ninguna relacion de este tipo en la tabla

## 5.5 Expandir datos

In [63]:
# perfil_ingreso_v2_df['AÑO'], perfil_ingreso_v2_df['SEMESTRE'] = expandir_periodo(perfil_ingreso_v2_df.PERIODO_COHORTE)
# perfil_ingreso_v2_df['FECHA'] = fecha_de_semestre(perfil_ingreso_v2_df['AÑO'], perfil_ingreso_v2_df['SEMESTRE'])

## 5.6 Consistencia en los datos

In [64]:
perfil_ingreso_v2_df.COD_MEN.str.len().value_counts()

5    9474
6     926
Name: COD_MEN, dtype: int64

> tambiens se puede normalizar el numero de digitos

In [65]:
# cols_to_fill = ['PERIODO_COHORTE', 'CODIGO', 'ULTIMA_MATRICULA', 'COD_MEN']
# filled = perfil_ingreso_v2_df[cols_to_fill].apply(fill_digits, index=0)
# perfil_ingreso_v2_df[cols_to_fill] = filled

## 5.7 Resultado

In [66]:
perfil_ingreso_v2_df.head()

PERIODO_COHORTE    CODIGO GENERO  EDAD  \
0          201210  11312001      M    22   
1          201210  11312002      M    23   
2          201210  11312003      F    22   
3          201210  11312004      F    25   
4          201210  11312005      M    25   

                                 COLEGIO_PROCEDENCIA  PUNT_TOTAL  PUNT_AREA  \
0                              col coop cooservicios         438         49   
1  instituto de educacion media diversificada ine...         386         48   
2                                i.e. antonio narino         460         51   
3                         i.e. luis hernandez vargas          41         42   
4                     ie tec jose ignacio de marquez          50         49   

  ES_DESERTOR ULTIMA_MATRICULA COD_MEN  
0          no           201710   20708  
1          no           201420   20708  
2          no           201620   20708  
3          no           201620   20708  
4          no           201610   20708

In [67]:
perfil_ingreso_v2_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10400 entries, 0 to 10399
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   PERIODO_COHORTE      10400 non-null  object
 1   CODIGO               10400 non-null  object
 2   GENERO               10400 non-null  object
 3   EDAD                 10400 non-null  int32 
 4   COLEGIO_PROCEDENCIA  10400 non-null  object
 5   PUNT_TOTAL           10400 non-null  int32 
 6   PUNT_AREA            10400 non-null  int32 
 7   ES_DESERTOR          10400 non-null  object
 8   ULTIMA_MATRICULA     10400 non-null  object
 9   COD_MEN              10400 non-null  object
dtypes: int32(3), object(7)
memory usage: 690.8+ KB


## 6. Tabla "rendimiento_academico"

In [68]:
rendimiento_academico_df.head()

PERIODO_COHORTE  PERIODO_ACADEMICO    CODIGO  COD_MAT           NOMBRE_MAT  \
0           201210             201210  11312001    11305    EXP. ORAL ESCRITA   
1           201210             201210  11312001    11312    INICIATIVA EMPRES   
2           201210             201210  11312001    11310        MICROECONOMÍA   
3           201210             201210  11312001    11315   INFORMÁTICA BÁSICA   
4           201210             201210  11312001    11303  INTROD. AL PROGRAMA   

  NOTA_DEF    ESTADO  
0     3,50  APROBADA  
1     4,00  APROBADA  
2     3,60  APROBADA  
3     3,30  APROBADA  
4     3,80  APROBADA

In [69]:
rendimiento_academico_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 364492 entries, 0 to 364491
Data columns (total 7 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   PERIODO_COHORTE    364492 non-null  int64 
 1   PERIODO_ACADEMICO  364492 non-null  int64 
 2   CODIGO             364492 non-null  int64 
 3   COD_MAT            364492 non-null  int64 
 4   NOMBRE_MAT         364492 non-null  object
 5   NOTA_DEF           364492 non-null  object
 6   ESTADO             364492 non-null  object
dtypes: int64(4), object(3)
memory usage: 19.5+ MB


In [70]:
rendimiento_academico_df.nunique()

PERIODO_COHORTE         19
PERIODO_ACADEMICO       35
CODIGO               10211
COD_MAT               1839
NOMBRE_MAT             913
NOTA_DEF                51
ESTADO                   2
dtype: int64

> * No hay valores nulos
> * "NOMBRE_MAT" Texto a normalizar
> * Convertir variables categoricas de periodo y codigos a string
> * "NOTA" debe ser float
> * "COD_MAT" y "NOMBRE_MAT" no coinciden en cantidad

## 6.1 Llenar valores faltantes

> No hay valores faltantes

## 6.2 Corregir tipo de datos

In [71]:
to_str_cols = ['PERIODO_COHORTE', 'PERIODO_ACADEMICO', 'CODIGO', 'COD_MAT']
rendimiento_academico_df[to_str_cols] = rendimiento_academico_df[to_str_cols].astype(str)

rendimiento_academico_df['NOTA_DEF'] = rendimiento_academico_df['NOTA_DEF'].str.replace(',','.').astype(float)

## 6.3 Normalizar Texto

In [72]:
rendimiento_academico_df.NOMBRE_MAT = normalize_text(rendimiento_academico_df.NOMBRE_MAT)
rendimiento_academico_df.ESTADO = normalize_text(rendimiento_academico_df.ESTADO)

## 6.4 Revisar llaves primarias de categorias

Esta la relacion "COD_MAT" - "NOMBRE_MAT"

> * Hay mas codigos que nombres
> * Existen multiples variaciones de nombre para el mismo codigo: se deberian unificar
> * para el mismo nombre pueden existir multiples codigos: puede ser debido a que es la misma materia o similar, pero de otro porgrama, facultad o año. NO se deberia unificar.
> * se puede dejar que

Puedo empezar tomando para un mismo codigo un solo nombre (primero, max, min, string comun o la suma de los strings).

luego para mismos nombres tomar el menor codigo

In [73]:
# multiples nombres para el mismo codigo:
codigos = rendimiento_academico_df.groupby(['COD_MAT', 'NOMBRE_MAT'])[['ESTADO']].count().reset_index() # count # filas para cada combiancion cod-nombre
agg_func_text = {'NOMBRE_MAT': {'nunique', set, 'count'}} # count # veces codigo aparece (# de combinaciones)
codigos2 = codigos.groupby('COD_MAT').agg(agg_func_text)
dict(codigos2[codigos2['NOMBRE_MAT']['count'] > 1]['NOMBRE_MAT']['set']) # > 1, mas de una combinacion (codigos que poseen mas de un nombre)

{'11317': {'idioma extranj. i', 'idioma extranjero i', 'ingles i'},
 '11330': {'idioma extranjero ii', 'ingles ii'},
 '11401': {'matemat basica', 'matematica basica'},
 '11403': {'introd al  program', 'introd. al programa'},
 '11405': {'exp. oral escrita', 'expre oral escrita'},
 '11406': {'deport formativo', 'deporte formativo'},
 '11407': {'formac integ  i', 'formacion integ i'},
 '11410': {'contabil. basica', 'contabilidad'},
 '11411': {'estadistica', 'estadistica descrip.'},
 '11412': {'legilac c/cial', 'legislac.comercia'},
 '11413': {'admon basica', 'teorias de la admon'},
 '11414': {'calculo', 'calculo diferenc.'},
 '11416': {'cont financ i', 'contabil. activos'},
 '11417': {'idioma extranj.i', 'idioma extranjero i', 'ingles i'},
 '11418': {'idioma extranj.ii', 'idioma extranjero ii', 'ingles ii'},
 '11420': {'formac integ  ii', 'formacion integ ii'},
 '11422': {'elect libre', 'electiva libre i'},
 '11424': {'cont.costos.prod', 'contabil costos'},
 '11429': {'const form ciudadan

In [74]:
# multiples codigos para mismo nombre
codigos = rendimiento_academico_df.groupby(['COD_MAT', 'NOMBRE_MAT'])[['ESTADO']].count().reset_index()
codigos.NOMBRE_MAT.value_counts().head(15)

logica                  31
humanidades             30
electiva libre ii       29
ingles ii               29
ingles i                29
deporte formativo       28
formacion integ i       28
informatica basica      28
elect.compleme. ii      27
electiva libre i        27
formacion integ ii      27
introd. al programa     27
idioma extranjero ii    27
etica general           27
elect.profundi. ii      27
Name: NOMBRE_MAT, dtype: int64

In [75]:
codigos[codigos.NOMBRE_MAT=='logica']

COD_MAT NOMBRE_MAT  ESTADO
1      11302     logica     144
65     11402     logica     129
166    11702     logica     428
269    11802     logica     243
338    11902     logica      97
416    33102     logica     309
481    33202     logica     331
544    33302     logica     700
609    33402     logica      32
673    55102     logica     256
737    55202     logica     219
825    55302     logica     486
902    55402     logica     156
965    55502     logica     747
1032   55602     logica     333
1105   55702     logica     703
1176   55806     logica      10
1201   55906     logica       1
1217   66102     logica     616
1280   66202     logica     257
1347   66302     logica     243
1417   66402     logica    1795
1478   66502     logica     420
1547   66606     logica     928
1594   77102     logica     397
1684   77202     logica     366
1791   77308     logica     107
1847   77402     logica     288
1940   88102     logica     435
2002   88202     logica     359
2075   88313     logica      67

> NO SE A SOLUCIONADO

## 6.5 Expandir datos

In [76]:
# rendimiento_academico_df['AÑO'], rendimiento_academico_df['SEMESTRE'] = expandir_periodo(rendimiento_academico_df.PERIODO_ACADEMICO)
# rendimiento_academico_df['FECHA'] = fecha_de_semestre(rendimiento_academico_df['AÑO'], rendimiento_academico_df['SEMESTRE'])

## 6.6 Consistencia en los datos

In [77]:
# cols_to_fill = ['PERIODO_COHORTE', 'PERIODO_ACADEMICO', 'CODIGO', 'COD_MAT']
# filled = rendimiento_academico_df[cols_to_fill].apply(fill_digits, index=0)
# rendimiento_academico_df[cols_to_fill] = filled

## 6.7 Resultado

In [78]:
rendimiento_academico_df.head()

PERIODO_COHORTE PERIODO_ACADEMICO    CODIGO COD_MAT           NOMBRE_MAT  \
0          201210            201210  11312001   11305    exp. oral escrita   
1          201210            201210  11312001   11312    iniciativa empres   
2          201210            201210  11312001   11310        microeconomia   
3          201210            201210  11312001   11315   informatica basica   
4          201210            201210  11312001   11303  introd. al programa   

   NOTA_DEF    ESTADO  
0       3.5  aprobada  
1       4.0  aprobada  
2       3.6  aprobada  
3       3.3  aprobada  
4       3.8  aprobada

In [79]:
rendimiento_academico_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 364492 entries, 0 to 364491
Data columns (total 7 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   PERIODO_COHORTE    364492 non-null  object 
 1   PERIODO_ACADEMICO  364492 non-null  object 
 2   CODIGO             364492 non-null  object 
 3   COD_MAT            364492 non-null  object 
 4   NOMBRE_MAT         364492 non-null  object 
 5   NOTA_DEF           364492 non-null  float64
 6   ESTADO             364492 non-null  object 
dtypes: float64(1), object(6)
memory usage: 19.5+ MB


In [80]:
rendimiento_academico_df.nunique()

PERIODO_COHORTE         19
PERIODO_ACADEMICO       35
CODIGO               10211
COD_MAT               1839
NOMBRE_MAT             906
NOTA_DEF                51
ESTADO                   2
dtype: int64

## 7. Tabla "area_icfes"

In [81]:
area_icfes_df.head()

FECHA_INI_EX FECHA_FIN_EX  COD_MEN          AREA_ICFES
0   1980-05-01   1999-12-31    20697  CIENCIAS NATURALES
1   1980-05-01   1999-12-31    20697            LENGUAJE
2   1980-05-01   1999-12-31    20697          MATEMÁTICA
3   2000-01-01   2004-12-24    20697            BIOLOGÍA
4   2000-01-01   2004-12-24    20697              FÍSICA

In [82]:
area_icfes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 674 entries, 0 to 673
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   FECHA_INI_EX  674 non-null    datetime64[ns]
 1   FECHA_FIN_EX  674 non-null    datetime64[ns]
 2   COD_MEN       674 non-null    int64         
 3   AREA_ICFES    674 non-null    object        
dtypes: datetime64[ns](2), int64(1), object(1)
memory usage: 21.2+ KB


In [83]:
area_icfes_df.nunique()

FECHA_INI_EX     8
FECHA_FIN_EX     8
COD_MEN         23
AREA_ICFES      16
dtype: int64

> * normalizar texto de las areas
> * "COD_MEN" categorico, convertir a string
> * las fechas no son relevantes para nuestro analisis
> * hay mas valores unicos de "COD_MEN" que de "AREA_ICFES"

## 7.1 Llenar valores faltantes

> * NO hay valores faltantes

## 7.2 Corregir tipo de datos

In [84]:
area_icfes_df["COD_MEN"] = area_icfes_df["COD_MEN"].astype(str)

## 7.3 Normalizar Texto

In [85]:
area_icfes_df.AREA_ICFES = normalize_text(area_icfes_df.AREA_ICFES)

## 7.4 Revisar llaves primarias de categorias

In [86]:

agg_func_text = {'AREA_ICFES': {'nunique', set, 'count'}}
area_icfes_df.groupby(['COD_MEN', 'AREA_ICFES']).count().reset_index().groupby('COD_MEN').agg(agg_func_text)['AREA_ICFES']['set'].head()

COD_MEN
101677    {fisica, cienc.soc.y ciudadan, sociales, quimi...
105810    {historia, cienc.soc.y ciudadan, sociales, fil...
108452    {fisica, ciencias naturales, lectura critica, ...
108951    {fisica, cienc.soc.y ciudadan, soc. y ciudadan...
20697     {fisica, cienc.soc.y ciudadan, quimica, cienci...
Name: set, dtype: object

> Contrario a lo ques se "COD_MEN" no es una llave para "AREA_ICFES", entonces no se hacen cambios

## 7.5 Expandir datos

> * no hay que expandir

## 7.6 Consistencia en los datos

In [87]:
# df['COD_MEN'] = fill_digits(df['COD_MEN'])

## 7.7 Resultado

In [88]:
area_icfes_df.head()

FECHA_INI_EX FECHA_FIN_EX COD_MEN          AREA_ICFES
0   1980-05-01   1999-12-31   20697  ciencias naturales
1   1980-05-01   1999-12-31   20697            lenguaje
2   1980-05-01   1999-12-31   20697          matematica
3   2000-01-01   2004-12-24   20697            biologia
4   2000-01-01   2004-12-24   20697              fisica

In [89]:
area_icfes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 674 entries, 0 to 673
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   FECHA_INI_EX  674 non-null    datetime64[ns]
 1   FECHA_FIN_EX  674 non-null    datetime64[ns]
 2   COD_MEN       674 non-null    object        
 3   AREA_ICFES    674 non-null    object        
dtypes: datetime64[ns](2), object(2)
memory usage: 21.2+ KB


In [90]:
area_icfes_df.nunique()

FECHA_INI_EX     8
FECHA_FIN_EX     8
COD_MEN         23
AREA_ICFES      16
dtype: int64

## 8. Tabla "creditos_periodo"

In [91]:
creditos_periodo_df.head()

CODIGO  PERIODO
0  33111541   201210
1  66411591   201210
2  33103565   201210
3  66411104   201210
4  66409105   201210

In [92]:
creditos_periodo_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7998 entries, 0 to 7997
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   CODIGO   7998 non-null   int64
 1   PERIODO  7998 non-null   int64
dtypes: int64(2)
memory usage: 125.1 KB


In [93]:
creditos_periodo_df.nunique()

CODIGO     2707
PERIODO      20
dtype: int64

> * Tabla parece incompleta
> * No hay valores faltantes
> * convertir datos a strings

## 8.1 Llenar valores faltantes

> No hay valores faltantes

## 8.2 Corregir tipo de datos

In [94]:
to_str_cols = ['CODIGO', 'PERIODO']
creditos_periodo_df[to_str_cols] = creditos_periodo_df[to_str_cols].astype(str)

# to_int_cols = []
# creditos_periodo_df[to_int_cols] = creditos_periodo_df[to_int_cols].astype(int)

# to_float_cols = []
# creditos_periodo_df[to_float_cols] = creditos_periodo_df[to_float_cols].astype(float)

## 8.3 Normalizar Texto

> No hay columnas de textom

## 8.4 Revisar llaves primarias de categorias

No hay ninguna relacion de este tipo

## 8.5 Expandir datos

In [95]:
# creditos_periodo_df['AÑO'], creditos_periodo_df['SEMESTRE'] = expandir_periodo(creditos_periodo_df.PERIODO)
# creditos_periodo_df['FECHA'] = fecha_de_semestre(creditos_periodo_df['AÑO'], creditos_periodo_df['SEMESTRE'])

## 8.6 Consistencia en los datos

In [96]:
# cols_to_fill = ['CODIGO', 'PERIODO']
# filled = creditos_periodo_df[cols_to_fill].apply(fill_digits, index=0)
# creditos_periodo_df[cols_to_fill] = filled

## 8.7 Resultado

In [97]:
creditos_periodo_df.head()

CODIGO PERIODO
0  33111541  201210
1  66411591  201210
2  33103565  201210
3  66411104  201210
4  66409105  201210

In [98]:
creditos_periodo_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7998 entries, 0 to 7997
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   CODIGO   7998 non-null   object
 1   PERIODO  7998 non-null   object
dtypes: object(2)
memory usage: 125.1+ KB


In [99]:
creditos_periodo_df.nunique()

CODIGO     2707
PERIODO      20
dtype: int64

## 9. Tabla "materias_por_programa"

In [100]:
materias_por_programa_df.head()

COD_MEN                                          FACULTAD  \
0    20708  FACULTAD DE CIENCIAS ADMINISTRATIVAS Y CONTABLES   
1    20708  FACULTAD DE CIENCIAS ADMINISTRATIVAS Y CONTABLES   
2    20708  FACULTAD DE CIENCIAS ADMINISTRATIVAS Y CONTABLES   
3    20708  FACULTAD DE CIENCIAS ADMINISTRATIVAS Y CONTABLES   
4    20708  FACULTAD DE CIENCIAS ADMINISTRATIVAS Y CONTABLES   

              NOMBRE_PROGRAMA ACREDITADO        SEDE  COD_MAT  NUM_CREDITOS  \
0  ADMINISTRACIÓN DE EMPRESAS       N.A.  SEDE TUNJA    11330             2   
1  ADMINISTRACIÓN DE EMPRESAS       N.A.  SEDE TUNJA    11329             2   
2  ADMINISTRACIÓN DE EMPRESAS       N.A.  SEDE TUNJA    11334             2   
3  ADMINISTRACIÓN DE EMPRESAS       N.A.  SEDE TUNJA    11335             3   
4  ADMINISTRACIÓN DE EMPRESAS       N.A.  SEDE TUNJA    11336             2   

  INTRODUCCION  
0           NO  
1           NO  
2           NO  
3           NO  
4           NO

In [101]:
materias_por_programa_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1996 entries, 0 to 1995
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   COD_MEN          1996 non-null   int64 
 1   FACULTAD         1996 non-null   object
 2   NOMBRE_PROGRAMA  1996 non-null   object
 3   ACREDITADO       1996 non-null   object
 4   SEDE             1996 non-null   object
 5   COD_MAT          1996 non-null   int64 
 6   NUM_CREDITOS     1996 non-null   int64 
 7   INTRODUCCION     1996 non-null   object
dtypes: int64(3), object(5)
memory usage: 124.9+ KB


In [102]:
materias_por_programa_df.nunique()

COD_MEN              25
FACULTAD              6
NOMBRE_PROGRAMA      31
ACREDITADO            3
SEDE                  2
COD_MAT            1875
NUM_CREDITOS         16
INTRODUCCION          2
dtype: int64

> * Normalizar Stirngs
> * No hay valores nulos

## 9.1 Llenar valores faltantes

> No hay valores faltantes

## 9.2 Corregir tipo de datos

In [103]:
to_str_cols = ['COD_MEN', 'COD_MAT']
materias_por_programa_df[to_str_cols] = materias_por_programa_df[to_str_cols].astype(str)

to_int_cols = []
materias_por_programa_df[to_int_cols] = materias_por_programa_df[to_int_cols].astype(int)

to_float_cols = []
materias_por_programa_df[to_float_cols] = materias_por_programa_df[to_float_cols].astype(float)

## 9.3 Normalizar Texto

In [104]:
cols = ['FACULTAD', 'NOMBRE_PROGRAMA', 'ACREDITADO', 'SEDE', 'INTRODUCCION']
materias_por_programa_df[cols] = materias_por_programa_df[cols].apply(normalize_text, axis=0)

materias_por_programa_df.SEDE = materias_por_programa_df.SEDE.str.replace('sede ', '')

## 9.4 Revisar llaves primarias de categorias

> * No hay relaciones de este tipo
> * sigo sin saber que es "COD_MEN"

## 9.5 Expandir datos

> Podemos expandir acreditados en: 
> * acrediatdo nacional (SI, NO)
> * acreditado internacional (SI, NO)

In [105]:
materias_por_programa_df.ACREDITADO.value_counts()

n.a.        886
cna-riev    874
riev        236
Name: ACREDITADO, dtype: int64

> * cna: consejo naconal de acreditacion
> * riev: red internacional de evaluadores

In [106]:
materias_por_programa_df['ES_ACREDITADO'] = materias_por_programa_df.ACREDITADO.str.contains('n.a').astype(int).map({1:'no', 0:'si'})
materias_por_programa_df['ACRED_NACIONAL'] = materias_por_programa_df.ACREDITADO.str.contains('cna').astype(int).map({1:'si', 0:'no'})
materias_por_programa_df['ACRED_INTERNACIONAL'] = materias_por_programa_df.ACREDITADO.str.contains('riev').astype(int).map({1:'si', 0:'no'})

## 9.6 Consistencia en los datos

In [107]:
# cols_to_fill = ['COD_MEN', 'COD_MAT']
# filled = materias_por_programa_df[cols_to_fill].apply(fill_digits, index=0)
# materias_por_programa_df[cols_to_fill] = filled

## 9.7 Resultado

In [108]:
materias_por_programa_df.head()

COD_MEN                                          FACULTAD  \
0   20708  facultad de ciencias administrativas y contables   
1   20708  facultad de ciencias administrativas y contables   
2   20708  facultad de ciencias administrativas y contables   
3   20708  facultad de ciencias administrativas y contables   
4   20708  facultad de ciencias administrativas y contables   

              NOMBRE_PROGRAMA ACREDITADO   SEDE COD_MAT  NUM_CREDITOS  \
0  administracion de empresas       n.a.  tunja   11330             2   
1  administracion de empresas       n.a.  tunja   11329             2   
2  administracion de empresas       n.a.  tunja   11334             2   
3  administracion de empresas       n.a.  tunja   11335             3   
4  administracion de empresas       n.a.  tunja   11336             2   

  INTRODUCCION ES_ACREDITADO ACRED_NACIONAL ACRED_INTERNACIONAL  
0           no            no             no                  no  
1           no            no             no                  no  
2           no            no             no                  no  
3           no            no             no                  no  
4           no            no             no                  no

In [109]:
materias_por_programa_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1996 entries, 0 to 1995
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   COD_MEN              1996 non-null   object
 1   FACULTAD             1996 non-null   object
 2   NOMBRE_PROGRAMA      1996 non-null   object
 3   ACREDITADO           1996 non-null   object
 4   SEDE                 1996 non-null   object
 5   COD_MAT              1996 non-null   object
 6   NUM_CREDITOS         1996 non-null   int64 
 7   INTRODUCCION         1996 non-null   object
 8   ES_ACREDITADO        1996 non-null   object
 9   ACRED_NACIONAL       1996 non-null   object
 10  ACRED_INTERNACIONAL  1996 non-null   object
dtypes: int64(1), object(10)
memory usage: 171.7+ KB


In [110]:
materias_por_programa_df.nunique()

COD_MEN                  25
FACULTAD                  6
NOMBRE_PROGRAMA          31
ACREDITADO                3
SEDE                      2
COD_MAT                1875
NUM_CREDITOS             16
INTRODUCCION              2
ES_ACREDITADO             2
ACRED_NACIONAL            2
ACRED_INTERNACIONAL       2
dtype: int64

## 10. Tabla "pre_correquisito"

In [111]:
pre_correquisito_df.head()

COD_BASE              NOMBRE_MAT  COD_PRECO NOMBRE_PRECORREQUISITO  \
0     11304    PLANEACIÓN Y CONTROL      11313  ADMINISTRACIÓN BÁSICA   
1     11308     CÁLCULO DIFERENCIAL      11301      MATEMÁTICA BÁSICA   
2     11309  CONTABILIDAD COMERCIAL      11311    CONTABILIDAD BÁSICA   
3     11314    CREACIÓN DE EMPRESAS      11313  ADMINISTRACIÓN BÁSICA   
4     11316           MACROECONOMÍA      11310          MICROECONOMÍA   

  TIPO_REQUISITO  
0  PRERREQUISITO  
1  PRERREQUISITO  
2  PRERREQUISITO  
3  PRERREQUISITO  
4  PRERREQUISITO

In [112]:
pre_correquisito_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2709 entries, 0 to 2708
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   COD_BASE                2709 non-null   int64 
 1   NOMBRE_MAT              2709 non-null   object
 2   COD_PRECO               2709 non-null   int64 
 3   NOMBRE_PRECORREQUISITO  2709 non-null   object
 4   TIPO_REQUISITO          2709 non-null   object
dtypes: int64(2), object(3)
memory usage: 105.9+ KB


In [113]:
pre_correquisito_df.nunique()

COD_BASE                   819
NOMBRE_MAT                 433
COD_PRECO                 1526
NOMBRE_PRECORREQUISITO     703
TIPO_REQUISITO               2
dtype: int64

> * No hay valores faltantes
> * convertir variables "COD" categoricas a strings
> * Normalizar texto

## 10.1 Llenar valores faltantes

> No hay valores faltantes

## 10.2 Corregir tipo de datos

In [114]:
to_str_cols = ['COD_BASE', 'COD_PRECO']
pre_correquisito_df[to_str_cols] = pre_correquisito_df[to_str_cols].astype(str)

# to_int_cols = []
# pre_correquisito_df[to_int_cols] = pre_correquisito_df[to_int_cols].astype(int)

# to_float_cols = []
# pre_correquisito_df[to_float_cols] = pre_correquisito_df[to_float_cols].astype(float)

## 10.3 Normalizar Texto

In [115]:
cols = ['NOMBRE_MAT',	'NOMBRE_PRECORREQUISITO', 'TIPO_REQUISITO']
pre_correquisito_df[cols] = pre_correquisito_df[cols].apply(normalize_text, axis=0)

## 10.4 Revisar llaves primarias de categorias

> Revisar "COD_BASE"

## 10.5 Expandir datos

> No hay datos a expandir

## 10.6 Consistencia en los datos

In [116]:
# cols_to_fill = ['COD_BASE', 'COD_PRECO']
# filled = pre_correquisito_df[cols_to_fill].apply(fill_digits, index=0)
# pre_correquisito_df[cols_to_fill] = filled

## 10.7 Resultado

In [117]:
pre_correquisito_df.head()

COD_BASE              NOMBRE_MAT COD_PRECO NOMBRE_PRECORREQUISITO  \
0    11304    planeacion y control     11313  administracion basica   
1    11308     calculo diferencial     11301      matematica basica   
2    11309  contabilidad comercial     11311    contabilidad basica   
3    11314    creacion de empresas     11313  administracion basica   
4    11316           macroeconomia     11310          microeconomia   

  TIPO_REQUISITO  
0  prerrequisito  
1  prerrequisito  
2  prerrequisito  
3  prerrequisito  
4  prerrequisito

In [118]:
pre_correquisito_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2709 entries, 0 to 2708
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   COD_BASE                2709 non-null   object
 1   NOMBRE_MAT              2709 non-null   object
 2   COD_PRECO               2709 non-null   object
 3   NOMBRE_PRECORREQUISITO  2709 non-null   object
 4   TIPO_REQUISITO          2709 non-null   object
dtypes: object(5)
memory usage: 105.9+ KB


In [119]:
pre_correquisito_df.nunique()

COD_BASE                   819
NOMBRE_MAT                 432
COD_PRECO                 1526
NOMBRE_PRECORREQUISITO     699
TIPO_REQUISITO               2
dtype: int64

## 11. Tabla "reingreso_aplazamiento"

In [120]:
reingreso_aplazamiento_df.head()

CODIGO FECHA_SOLICITUD       TIPO
0  55110503      2012-01-10  REINGRESO
1  55510048      2012-01-10  REINGRESO
2  55510535      2012-01-10  REINGRESO
3  55511017      2012-01-10  REINGRESO
4  55511018      2012-01-10  REINGRESO

In [121]:
reingreso_aplazamiento_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6701 entries, 0 to 6700
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   CODIGO           6701 non-null   int64         
 1   FECHA_SOLICITUD  6701 non-null   datetime64[ns]
 2   TIPO             6701 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 157.2+ KB


In [122]:
reingreso_aplazamiento_df.nunique()

CODIGO             3891
FECHA_SOLICITUD    1606
TIPO                  2
dtype: int64

> * 

## 11.1 Llenar valores faltantes

> No hay valores Faltantes

## 11.2 Corregir tipo de datos

In [123]:
reingreso_aplazamiento_df['CODIGO'] = reingreso_aplazamiento_df['CODIGO'].astype(str)

# to_int_cols = []
# reingreso_aplazamiento_df[to_int_cols] = reingreso_aplazamiento_df[to_int_cols].astype(int)

# to_float_cols = []
# reingreso_aplazamiento_df[to_float_cols] = reingreso_aplazamiento_df[to_float_cols].astype(float)

## 11.3 Normalizar Texto

In [124]:
reingreso_aplazamiento_df.TIPO = normalize_text(reingreso_aplazamiento_df.TIPO)

### 11.4 Revisar llaves primarias de categorias

> No hay relaciones de este tipo

## 11.5 Expandir datos

> NO hay que expandir

## 11.6 Consistencia en los datos

In [125]:
# reingreso_aplazamiento_df['CODIGO'] = fill_digits(reingreso_aplazamiento_df['CODIGO'] )

## 11.7 Resultado

In [126]:
reingreso_aplazamiento_df.head()

CODIGO FECHA_SOLICITUD       TIPO
0  55110503      2012-01-10  reingreso
1  55510048      2012-01-10  reingreso
2  55510535      2012-01-10  reingreso
3  55511017      2012-01-10  reingreso
4  55511018      2012-01-10  reingreso

In [127]:
reingreso_aplazamiento_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6701 entries, 0 to 6700
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   CODIGO           6701 non-null   object        
 1   FECHA_SOLICITUD  6701 non-null   datetime64[ns]
 2   TIPO             6701 non-null   object        
dtypes: datetime64[ns](1), object(2)
memory usage: 157.2+ KB


In [128]:
reingreso_aplazamiento_df.nunique()

CODIGO             3891
FECHA_SOLICITUD    1606
TIPO                  2
dtype: int64

# Conclusiones

1. A las tablas originales no les faltan muchos datos y fueron facilmente reemplazados.
> * la mayoria de datos faltantes correspondian a ceros en creditos de materias inscirtas o aprobadas, para los casos que un estudiante no tenia materias inscritas en un semestre.
> * El unico caso que se tuvo que llenar arbitrariamente fue con la edad de estudiantes. Como solo faltaban dos datos (no es importante), se reemplazaron estos dos calores con la moda.
2. Se convieritio el tipo de dato a string, para todas las variables categoricas representadas por numeros. (CODIGOS de maestro, materia, estudiante, etc)
4. Se corrigieron los codigos que funcionaban como llaves primarias para variables categoricas.
> * Por ejemplo se hizo para el caso de Factores de evaluacion docente, los cuales en algunos casos tenian dos codigos para el mismo factor.
3. Se expandio las columnas de periodo en año, semestre y fecha (en caso de que pueda facilitar analisis futuros)
4. El proceso de limpieza extraido de este analisis se consolida de manera automatizad en el script **"src/data/clean_data.py"**

5. **Pendiente: unificar una nueva columna de materias (por nombre unico y no por codigo)**